src/Analyzers/Core/Analyzers/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs

```diff
@@ -43,274 +53,186 @@ protected AbstractUseAutoPropertyAnalyzer()
+                context.RegisterCodeBlockStartAction<TSyntaxKind>(context =>
+                    RegisterIneligibleFieldsAction(ineligibleFields, context.SemanticModel, context.CodeBlock, context.CancellationToken));
 
-            RegisterIneligibleFieldsAction(
-                analysisResults, ineligibleFields,
-                semanticModel.Compilation, cancellationToken);
-            Process(analysisResults, ineligibleFields, context);
-        }
+                context.RegisterSymbolEndAction(context =>
+                    Process(analysisResults, ineligibleFields, context));
+            }, SymbolKind.NamedType);
 
-        protected void AnalyzeProperty(
-            SemanticModelAnalysisContext context, TPropertyDeclaration propertyDeclaration, List<AnalysisResult> analysisResults)
+        private void AnalyzePropertyDeclaration(
+            SyntaxNodeAnalysisContext context,
+            INamedTypeSymbol containingType,
+            ConcurrentQueue<AnalysisResult> analysisResults)
         {
             var cancellationToken = context.CancellationToken;
             var semanticModel = context.SemanticModel;
+            var compilation = semanticModel.Compilation;
 
+            var propertyDeclaration = (TPropertyDeclaration)context.Node;
             if (semanticModel.GetDeclaredSymbol(propertyDeclaration, cancellationToken) is not IPropertySymbol property)
-            {
                 return;
-            }
 
             if (property.IsIndexer)
-            {
                 return;
-            }
 
             // The property can't be virtual.  We don't know if it is overridden somewhere.  If it 
             // is, then calls to it may not actually assign to the field.
             if (property.IsVirtual || property.IsOverride || property.IsSealed)
-            {
                 return;
-            }
 
             if (property.IsWithEvents)
-            {
                 return;
-            }
 
             if (property.Parameters.Length > 0)
-            {
                 return;
-            }
 
             // Need at least a getter.
             if (property.GetMethod == null)
-            {
                 return;
-            }
 
             if (!CanExplicitInterfaceImplementationsBeFixed() && property.ExplicitInterfaceImplementations.Length != 0)
-            {
-                return;
-            }
-
-            var containingType = property.ContainingType;
-            if (containingType == null)
-            {
                 return;
-            }
 
             // Serializable types can depend on fields (and their order).  Don't report these
             // properties in that case.
             if (containingType.IsSerializable)
-            {
                 return;
-            }
 
-            var declarations = property.DeclaringSyntaxReferences;
-            if (declarations.Length != 1)
-            {
+            var preferAutoProps = context.GetAnalyzerOptions().PreferAutoProperties;
```

Comment:
>var preferAutoProps = context.GetAnalyzerOptions().PreferAutoProperties; [](http://example.com/codeflow?start=12&length=72)

I assume this can't be checked before registering any work?

Reply:
Analyzer options are on a perf-file basis, but a type may spread over different files (each with different options).  So you have to check once you get down to actual syntax.  

That said, we could check up front per type if any of its parts is in a file that has this set.  if not, we can bial out.  i'll make that change.

Reply:
done.

Reply:
yes.  the first check just establishes that *one* file the type is in has this option on.  But the type may have parts that are in files where it is still off.  so we still need to check when we hit each property decl.